sw/device/silicon_creator/lib/drivers/kmac.c

```diff
@@ -0,0 +1,279 @@
+  // Set `CFG.ERR_PROCESSED` bit to 0.
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_ERR_PROCESSED_BIT, 0);
+  // Set `CFG.EN_UNSUPPORTED_MODESTRENGTH` bit to 0.
+  cfg_reg = bitfield_bit32_write(
+      cfg_reg, KMAC_CFG_SHADOWED_EN_UNSUPPORTED_MODESTRENGTH_BIT, 0);
+  abs_mmio_write32_shadowed(kBase + KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+
+  // Write entropy seed registers (all-zero).
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_0_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_1_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_2_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_3_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_4_REG_OFFSET, 0);
+
+  // Read back and double-check the security strength and mode.
+  cfg_reg = abs_mmio_read32(kBase + KMAC_CFG_SHADOWED_REG_OFFSET);
+  HARDENED_CHECK_EQ(
+      bitfield_field32_read(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD),
+      KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L256);
+  HARDENED_CHECK_EQ(
+      bitfield_field32_read(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD),
+      KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE);
+
+  return kErrorOk;
+}
+
+rom_error_t kmac_shake256_start(void) {
+  // Block until KMAC hardware is idle.
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_IDLE_BIT));
+
+  // Issue `CMD.START` to start the operation.
+  issue_command(KMAC_CMD_CMD_VALUE_START);
+
+  // Block until KMAC hardware is in the `absorb` state. After `CMD.START`,
+  // KMAC should never move out of the `absorb` state until `CMD.PROCESS` is
+  // issued, so we get significant performance gains by polling only once here
+  // instead of before every `absorb`.
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_ABSORB_BIT));
+
+  return kErrorOk;
+}
+
+inline void kmac_shake256_absorb(const uint8_t *in, size_t inlen) {
+  // This implementation does not poll `STATUS.fifo_depth` as recommended in
+  // the KMAC documentation. Normally, polling is required to prevent a
+  // deadlock scenario between Ibex, KMAC, and EDN. However, in this case it is
+  // safe to skip because `kmac_shake256_configure()` sets KMAC to use
+  // software-only entropy, and sets `kmac_en` to false (so KMAC will not
+  // produce entropy requests anyway). Since KMAC will therefore not block on
+  // EDN, it is guaranteed to keep processing message blocks. For more details,
+  // see the KMAC documentation:
+  //   https://docs.opentitan.org/hw/ip/kmac/doc/#fifo-depth-and-empty-status
+
+  // Use byte-wide writes until the input pointer is aligned.
+  // Note: writes to the KMAC message FIFO are not required to be aligned.
+  ptrdiff_t offset = misalignment32_of((uintptr_t)in);
+  size_t idx = 0;
+  if (offset != 0) {
+    size_t nbytes = alignof(uint32_t) - offset;
+    for (; idx < nbytes && idx < inlen; idx++) {
+      abs_mmio_write8(kBase + KMAC_MSG_FIFO_REG_OFFSET, in[idx]);
+    }
+  }
```

Comment:
nit: I think we can simplify these loops if we work directly on the buffer pointer and length.

```suggestion
  const char *data = buf;
  for (; len > 0 && (uintptr_t)data & 0x3; --len, ++data) {
    abs_mmio_write8(kBase + KMAC_MSG_FIFO_REG_OFFSET, *data);
  }
```